src/expr/impl/src/scalar/make_time.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use chrono::{NaiveDate, NaiveDateTime, NaiveTime};
+use risingwave_common::types::{Date, FloatExt, Time, Timestamp, Timestamptz, F64};
+use risingwave_expr::expr_context::TIME_ZONE;
+use risingwave_expr::{capture_context, function, ExprError, Result};
+
+use crate::scalar::timestamptz::timestamp_at_time_zone;
+
+pub fn make_naive_date(year: i32, month: i32, day: i32, for_timestamp: bool) -> Result<NaiveDate> {
+    if year == 0 {
+        return Err(ExprError::InvalidParam {
+            name: "year, month, day",
+            reason: format!("invalid date: {}-{}-{}", year, month, day).into(),
+        });
+    }
+    let adjusted_year = if year < 0 && !for_timestamp {
+        // For a year `-X`, The `Date` type is printed as `X+1 BC`, while `Timestamp` is printed as `-X`.
+        year + 1
+    } else {
+        year
+    };
```

Comment:
Did not quite get this part. Date / timestamp / timestamptz shall all follow the same rule:
* In PG SQL, the year before 1 is -1. There is no year 0. It is printed as `1 BC`.
* In ISO 8601 (chrono crate), the year before 1 is 0. It can be printed as `1 BC` (`year_ce` instead of `year`)

Practically it is okay to be incompatible with PG for BC years. But the logic here is both wrong and more complicated

In short, `year + 1` as long as `year < 0` when passing a PG SQL input to the chrono crate. This yields the correct value in memory during input. Whether showing `X+1 BC` or `-X` shall be adjusted on the output side.

Reply:
After discussion with xiangjin, we conduct `+1` for both `Date` and `Timestamp`/`Timestamptz`. However, as our `ToText` for `Date` and `Timestamp`/`Timestamptz` are different, the output of `make_date` and `make_timestamp/timestamptz` are different.

```
SELECT make_timestamp(-5, 02, 29, 08, 15, 55.33);
----
-0004-02-29 08:15:55.330

SELECT make_date(-5, 2, 29);
----
0005-02-29 BC
```

We'll fix it later in another PR by modify the impl of `ToText` for`Timestamp`/`Timestamptz`.